### ISSUE TITLE
TLS server only performs handshake after 'accept' is called

### ISSUE THREAD

**Issue by Kinnay** [agronholm/anyio#125]
Hi,

I would expect the following code to work:

```python
import anyio
import ssl

async def main():
	server_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
	server_context.load_cert_chain(certfile="cert.pem", keyfile="key.pem")
	
	client_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
	client_context.load_verify_locations(cafile="cert.pem")
	
	async with await anyio.create_tcp_server(1234, ssl_context=server_context) as server:
		async with await anyio.connect_tcp("localhost", 1234, ssl_context=client_context, autostart_tls=True) as client:
			print("connected") # This line is never reached
			conn = await server.accept()

anyio.run(main)
```

However, it gets stuck in `anyio.connect_tcp`. If I disable `autostart_tls` it connects fine but gets stuck when I call `start_tls` on the client. It seems like the server is only able to perform the TLS handshake if I call `accept` on the server object, as the following code works fine:

```python
import anyio
import ssl

async def task(server):
	conn = await server.accept()
	await conn.close()

async def main():
	server_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
	server_context.load_cert_chain(certfile="cert.pem", keyfile="key.pem")
	
	client_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
	client_context.load_verify_locations(cafile="cert.pem")
	
	async with await anyio.create_tcp_server(1234, ssl_context=server_context) as server:
		async with anyio.create_task_group() as tg:
			await tg.spawn(task, server)
			async with await anyio.connect_tcp("localhost", 1234, ssl_context=client_context, autostart_tls=True) as client:
				print("connected") # This works fine

anyio.run(main)
```

I don't think this is the intended behavior?

The behavior seems to be the same across all backends.

**Comment by agronholm** [comment]
the `connect_tcp()` call does the handshake after connecting, but if the server side has not initiated the handshake on its own end, how do you expect the handshake to complete on the client side?

**Comment by Kinnay** [comment]
Maybe I misunderstood the meaning of `accept()`. In my view, `accept` has nothing to do with the handshake; it just waits until a connection is ready and returns a `SocketStream`. For example, the TCP server always does the handshake in the background, even if I don't call accept. This works fine:

```python
import anyio

async def main():
	async with await anyio.create_tcp_server(1234) as server:
		async with await anyio.connect_tcp("localhost", 1234) as client:
			print("connected") # This works fine
			conn = await server.accept()

anyio.run(main)
```

So the behavior is inconsistent between regular TCP servers and TLS servers. That's where my surprise came from.

**Comment by agronholm** [comment]
The difference is that a low-level `connect()` call succeeds if the server socket is just in listening mode and the new connection fits in the backlog. There is no way to know if the server has actually accepted the connection or not. On the other hand, the TLS handshake cannot complete until both sides are actively using the connection.

Try running the following code:
```python
import socket

server = socket.socket()
server.bind(('127.0.0.1', 0))
server.listen()
client = socket.socket()
client.connect(server.getsockname())
```
You'll see that it completes instantly and the server socket didn't even have to `accept()`.

Does this clear the confusion?

**Comment by Kinnay** [comment]
I see. Isn't one of the goals of an abstraction (such as `SocketStreamServer`) that its user does not have to worry about implementation details though?

TCP also involves a handshake. The only difference is that the TCP handshake is performed by the OS, whereas the TLS handshake is implemented by someone else, right? Why shouldn't a TLS connection succeed if the server is listening and it fits in the backlog?

**Comment by agronholm** [comment]
> Why shouldn't a TLS connection succeed if the server is listening and it fits in the backlog?

Because you cannot know if the handshake will succeed before it's actually performed.

**Comment by Kinnay** [comment]
Yes I get that. What I was trying to say is that I don't see why the TCP handshake can be performed in the background but the TLS handshake can not.

With TCP you also don't know if the handshake will succeed until it's performed. For example, you might be trying to connect to a port that the server is not listening on. In that case the handshake will definitely not succeed, but without trying to perform the handshake there is no way to tell if it will work.

From my view, both TCP and TLS establish a connection in a similar way:
1. Client initiates handshake
2. Server responds to handshake
3. (more steps here, e.g. client sends its certificate)
4. Connection is ready

The difference is:
* The TCP server responds to the handshake immediately after it is initiated by the client (in the background, before `accept()` is called).
* The TLS server responds to the handshake only after `accept()` is called.

In my opinion this difference does not make a lot of sense. I don't see why the TLS handshake can not be performed in the background? Or does some part of the TLS protocol not allow such behavior? I'm not really familiar with it to be honest.

It's not a big issue though, as I can spawn a task that accepts connections in the background relatively easily.

**Comment by agronholm** [comment]
When `connect()` succeeds, the handshake has already succeeded, the server just hasn't started working on the new connection. By "in the background", do you mean it should run in a separate task? Where would the errors be handled then? In the next task that tries to send to or read from the stream?

Anyway, if we allowed the TLS handshake to happen in the background, the client code might hand off the connection to the another component in the program even though it's not clear if it works at all. To me, it seems counterituitive to have `connect_tcp()` succeed if, say, the server certificate is invalid. The fact that you can `accept()` after `connect_tcp()` in the same task is just an artifact of the implementation and not something I would recommend.

**Comment by Kinnay** [comment]
> Anyway, if we allowed the TLS handshake to happen in the background, the client code might hand off the connection to the another component in the program even though it's not clear if it works at all. To me, it seems counterituitive to have connect_tcp() succeed if, say, the server certificate is invalid.

Of course. `connect` should only succeed after the handshake has succeeded. I didn't mean to say that the *client* should do the handshake in the background. I meant that the *server* should do it in the background. That is how to OS deals with TCP connections as well:
1. It always responds to the handshake in the background, regardless of whether anyone is calling `accept()`. This allows clients to connect even if no one has called `accept()`.
2. Once the handshake has completed, it adds the connection to a list of incoming connections.
3. `accept()` is independent from the handshake. It just pops a connection from the list.

In my opinion it would make sense if the TLS server followed the same behavior.

> By "in the background", do you mean it should run in a separate task? Where would the errors be handled then? In the next task that tries to send to or read from the stream?

I'm not sure if the server should deal with errors at all. I'm pretty sure that the TCP server just discards the connection if the client sends an invalid packet during the handshake. The TLS server could do the same (perhaps log a warning or something). Alternatively, it could perhaps raise an exception in `accept` for every failed client connection?

> The fact that you can accept() after connect_tcp() in the same task is just an artifact of the implementation and not something I would recommend.

Sure, you normally don't do that. I ran into this issue when I was writing tests for a library that implements a protocol on top of TLS. To check if everything works, I wanted to start both a server and a client in the test case and exchange a few messages. I expected that I can call `accept` *after* establishing the connection, because that's how TCP is implemented in the OS as well.

It took me a while to realize that I can only establish a connection if I call `accept` simultaneously. There is no warning. It just gets stuck.

I also wonder if this can lead to other issues at the server side. `accept` only returns after the handshake has succeeded. But that means that nothing is calling `accept` while the server is accepting a TLS connection. Imagine that a client is trying to connect that is really slow (perhaps even on purpose). Is the server still able to perform handshakes with other clients while it is waiting for the slow client to send its client certificate? Or does this effectively cause a denial of service?

**Comment by agronholm** [comment]
> I also wonder if this can lead to other issues at the server side. accept only returns after the handshake has succeeded. But that means that nothing is calling accept while the server is accepting a TLS connection. Imagine that a client is trying to connect that is really slow (perhaps even on purpose). Is the server still able to perform handshakes with other clients while it is waiting for the slow client to send its client certificate? Or does this effectively cause a denial of service?

You raise a good point here. I was about to refactor the serving/listening system to work more like trio anyway, so I'll make sure that `accept()` is only called in a dedicated task.

**Comment by agronholm** [comment]
In order to properly handle this, I'm going to have to go further than trio in the listener abstraction and give each listener more responsibility. I think I can also use this to avoid having to return multiple listeners from `open_tcp_listeners()`.

**Comment by njsmith** [comment]
Trio's `SSLListener.accept` always returns streams that *haven't* had their handshake performed. This is a major reason for why trio's `SSLStream` has an automatic-handshake-on-first-use feature – so if servers don't care about the handshake, they can just start using the stream and the handshake happens implicitly, or if they do care about the handshake, they can request it explicitly by calling `do_handshake`. Either way, though, the handshake always takes place in the context of the connection-specific task running the user's code, which is critical so that the user has a chance to do things like impose timeouts on the handshake process.

**Comment by agronholm** [comment]
Correct me if I'm wrong, but this "lazy" approach of doing the TLS handshake has also lead to major difficulties in the design of trio's SSL listener, hasn't it? The code looks **very** complex. Having a timeout for the handshake sounds like a good idea, but if the developer wants more control over the process, wrapping the stream with TLS manually will be trivial.

**Comment by njsmith** [comment]
> Correct me if I'm wrong, but this "lazy" approach of doing the TLS handshake has also lead to major difficulties in the design of trio's SSL listener, hasn't it? The code looks **very** complex.

It's two lines long... are we looking at the same code?

https://github.com/python-trio/trio/blob/e5bbc57ca6ef80be3422e6c05822dc04d80106c4/trio/_ssl.py#L911-L923

> Having a timeout for the handshake sounds like a good idea, but if the developer wants more control over the process, wrapping the stream with TLS manually will be trivial.

I guess my feeling is that every production use case has to have some kind of timeout that encompasses the handshake, so doing the handshake inside the listener would mean that the listener code is only suitable for demos.

**Comment by agronholm** [comment]
> It's two lines long... are we looking at the same code?

I'm talking about the lazy TLS negotiation. The code you pointed to does not do that.
I was looking at the `SSLStream` code.

> I guess my feeling is that every production use case has to have some kind of timeout that encompasses the handshake, so doing the handshake inside the listener would mean that the listener code is only suitable for demos.

With plain stdlib ssl sockets, you would set a socket timeout and then do a handshake, yes? Surely you're not saying that those too are only suitable for demos? Or maybe we are miscommunicating? My intent was to have the TLS listener accept a timeout value in the initializer, and anything more complex than that would mean the developer would have to do the wrapping manually.

**Comment by smurfix** [comment]
There's a difference between a standard 3-way TCP handshake (handled completely in kernel space, almost no impact unless you're a DDoS victim) and the TLS listener (userspace, complex, takes up non-negligible resources and CPU).

So yes if it's "only" a design choice then the listener task should own the negotiation.

Yes the code looks complex, but it's also correct and, what's more, it is a prime example of how to run SSL over any stream whatsoever. I'm somewhat less interested in a TLS implementation that's integrated into the "standard" kernel-based networking framework in such a way that if you ever need to run TLS over something else you need another implementation.

**Comment by agronholm** [comment]
> Yes the code looks complex, but it's also correct and, what's more, it is a prime example of how to run SSL over any stream whatsoever. I'm somewhat less interested in a TLS implementation that's integrated into the "standard" kernel-based networking framework in such a way that if you ever need to run TLS over something else you need another implementation.

I have a huge number of changes sitting in my local repository which, among many other things, decouple the TLS implementation from any socket functionality whatsoever and let you negotiate TLS over any arbitrary receive/send stream pair, just like trio does. These changes are very hard to separate from each other which is why I have elected not to push to master until I have much broader test coverage – something I'm actively working on.
